metsim/io.py

```diff
@@ -0,0 +1,227 @@
+"""
+IO Module for MetSim
+"""
+# Meteorology Simulator
+# Copyright (C) 2017  The Computational Hydrology Group, Department of Civil
+# and Environmental Engineering, University of Washington.
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>
+
+import os
+import struct
+
+import numpy as np
+import pandas as pd
+import xarray as xr
+
+from metsim.datetime import date_range
+
+
+def read_met_data(params: dict, domain: xr.Dataset) -> xr.Dataset:
+    """
+    Read input meteorological forcings for MetSim.
+    This method supports ascii, binary, netcdf, and
+    xarray input pointers.  The input source is derived
+    from the key 'forcing' in the params dictionary.
+    The format of the data is derived from 'in_format'
+    key in the parameter dictionary.
+    """
+    process_funcs = {
+        "netcdf": process_nc,
+        "binary": process_vic,
+        "ascii": process_vic,
+        "data": process_nc
+    }
+    return process_funcs[params['forcing_fmt']](params, domain)
+
+
+def read_domain(params: dict) -> xr.Dataset:
+    """Load in a domain file"""
+    read_funcs = {
+        "netcdf": read_netcdf,
+        "data": read_data
+    }
+    return read_funcs[params['domain_fmt']](
+        params['domain'], calendar=params['calendar'],
+        var_dict=params.get('domain_vars', None))
+
+
+def read_state(params: dict, domain: xr.Dataset) -> xr.Dataset:
+    """Load in a state file"""
+    read_funcs = {
+        "netcdf": read_netcdf,
+        "data": read_data
+    }
+    start = params['start'] - pd.Timedelta('90 days')
+    stop = params['start'] - pd.Timedelta('1 days')
+    return read_funcs[params['state_fmt']](
+        params['state'], domain=domain, iter_dims=params['iter_dims'],
+        start=start, stop=stop, calendar=params['calendar'],
+        var_dict=params.get('state_vars', None))
+
+
+def process_nc(params: dict, domain: xr.Dataset) -> xr.Dataset:
+    """Process NetCDF-like Data"""
+    read_funcs = {
+        "netcdf": read_netcdf,
+        "data": read_data
+    }
+    return read_funcs[params['forcing_fmt']](
+        params['forcing'], domain=domain, iter_dims=params['iter_dims'],
+        start=params['start'], stop=params['stop'],
+        calendar=params['calendar'], var_dict=params.get('forcing_vars', None))
+
+
+def process_vic(params: dict, domain: xr.Dataset) -> xr.Dataset:
```

Comment:
Is this VIC-ASCII or VIC-NetCDF? The naming is ambiguous, since the other `process_x` functions are specific to the format. If it is VIC-ASCII simply rename to `process_vic_ascii`